src/jit/importer.cpp

```diff
@@ -2207,6 +2207,13 @@ bool Compiler::impSpillStackEntry(unsigned level,
     /* Assign the spilled entry to the temp */
     impAssignTempGen(tnum, tree, verCurrentState.esStack[level].seTypeInfo.GetClassHandle(), level);
 
+    // If temp is single def and a ref type, grab what type info we can.
+    if (lvaTable[tnum].lvIsTemp && (lvaTable[tnum].lvType == TYP_REF))
+    {
+        CORINFO_CLASS_HANDLE stkHnd = verCurrentState.esStack[level].seTypeInfo.GetClassHandle();
+        lvaSetClass(tnum, tree, stkHnd);
+    }
```

Comment:
That case that I pointed out may be of interest to your devirtualization work.
If we have many return blocks (5 or more) we create a new LclVar to hold the return value and 
each return block gets converted into an assignment of the return value into this new LclVar and a branch to the "single" return block.   Thus we end up with a new LclVar that have 5 or more defs and a single use.  It has a "short-lifetime" as the JIT can typically find a register usually RAX to hold the value assigned and immediately returned.  If you were to treat this as having a single def you might infer that the return type is known when in fact there could be different return types.
This would probably only impact the inlining case and we probably don't try to inline a method with so many return blocks, but I'm not sure about that.